tests/unit/s2n_ktls_test_utils_test.c

```diff
@@ -0,0 +1,579 @@
+                total_sent += bytes_written;
+            }
+
+            /* confirm sent data */
+            EXPECT_OK(s2n_test_validate_data(&io_pair.client_in, test_data, total_sent));
+            /* validate `records_to_send` records were sent  */
+            EXPECT_EQUAL(s2n_stuffer_data_available(&io_pair.client_in.ancillary_buffer), records_to_send * S2N_TEST_KTLS_MOCK_HEADER_SIZE);
+            /* validate ancillary header */
+            for (size_t i = 0; i < records_to_send; i++) {
+                EXPECT_OK(s2n_test_validate_ancillary(&io_pair.client_in, test_record_type, S2N_TEST_TO_SEND));
+                /* consume the header in order to then validate the next header */
+                EXPECT_NOT_NULL(s2n_stuffer_raw_read(&io_pair.client_in.ancillary_buffer, S2N_TEST_KTLS_MOCK_HEADER_SIZE));
+            }
+
+            EXPECT_EQUAL(io_pair.client_in.send_recv_msg_invoked_count, records_to_send);
+        };
+
+        /* Send multiple records of different types */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_test_ktls_io_stuffer_pair io_pair = { 0 },
+                    s2n_ktls_io_stuffer_pair_free);
+            EXPECT_OK(s2n_test_init_ktls_io_stuffer(server, client, &io_pair));
+
+            size_t records_to_send = 5;
+            struct iovec send_msg_iov = { .iov_len = S2N_TEST_TO_SEND };
+            struct msghdr send_msg = { .msg_iov = &send_msg_iov, .msg_iovlen = 1 };
+            size_t total_sent = 0;
+            for (size_t i = 0; i < records_to_send; i++) {
+                /* increment test data ptr */
+                send_msg_iov.iov_base = test_data + total_sent;
+                send_msg.msg_control = &i;
+
+                /* sendmsg */
+                ssize_t bytes_written = s2n_test_ktls_sendmsg_io_stuffer(server->send_io_context, &send_msg);
+                EXPECT_EQUAL(bytes_written, S2N_TEST_TO_SEND);
+                total_sent += bytes_written;
+            }
+
+            /* confirm sent data */
+            EXPECT_OK(s2n_test_validate_data(&io_pair.client_in, test_data, total_sent));
+            /* validate `records_to_send` records were sent  */
+            EXPECT_EQUAL(s2n_stuffer_data_available(&io_pair.client_in.ancillary_buffer), records_to_send * S2N_TEST_KTLS_MOCK_HEADER_SIZE);
+            /* validate ancillary header */
+            for (size_t i = 0; i < records_to_send; i++) {
+                EXPECT_OK(s2n_test_validate_ancillary(&io_pair.client_in, i, S2N_TEST_TO_SEND));
+                /* consume the header in order to then validate the next header */
+                EXPECT_NOT_NULL(s2n_stuffer_raw_read(&io_pair.client_in.ancillary_buffer, S2N_TEST_KTLS_MOCK_HEADER_SIZE));
```

Comment:
If you're doing this after every call to s2n_test_validate_ancillary, that's a sign that I was wrong and you apparently DO want s2n_test_validate_ancillary to consume the stuffer :) I assumed you'd want to call this when you also planned to still recv with the mock.

Reply:
As I counted there are currently 6 total uses of validate_ancillary in tests. The 'consume' only happens in 2 of those tests (send only test where we send multiple records). If validate_ancillary consumed then we would potentially need to do rewind magic. We would also want to make the validate_data consume to make it match.

I like the idea of the validate functions being idempotent and non-mutating and then we can handle the small excepetion cases.